WeatherForecast [STEP3] Mason, Logan

qwerty3345 opened the pull request:
> 안녕하세요 라자냐~! 😆(@wonhee009) 최강 Team Melo의 Mason과 Logan입니다.!! 정말 오랜만이죠? 이래저래 고민할 것도 많고 프로젝트 말고도 할 일이 정말 많은 상황이라 프로젝트 진행속도가 생각보다 많이 느렸습니다. 그럼 이번에도 좋은 리뷰 부탁드려요!! 😊😊
> 

# ✅ 요구사항 & 체크리스트

- [x]  앱의 UI를 구현합니다
    - [x]  오토레이아웃을 활용하여 코드로만 구현
    - [x]  Collection View를 활용
    - [x]  Configuration을 활용
- [x]  화면을 위에서 아래로 끌었다 놓으면 정보를 새로고침합니다.

# 📲 동작 화면
![Simulator Screen Recording - iPhone 14 - 2023-04-06 at 15 14 23](https://user-images.githubusercontent.com/59835351/230288488-3c74957b-234e-4d23-8842-80ebb75e086c.gif)

# **⁉️ 고민과 해결**

## 💰 이미지 캐싱 구현

- 날씨 아이콘 이미지를 처음에는 assets 에 저장해뒀다가, 서버에서는 받아온 id 값과 매칭시켜서 업데이트를 해주려 했지만, 프로젝트 요구사항에 따라 이미지를 다운받는 식으로 변경하였습니다.
- 매번 모든 이미지를 다운받는 것의 단점을 느껴, 
NSCache를 활용한 이미지 캐시 매니저를 구현하고 이미지가 캐시에 존재한다면 캐싱을 통해 받아오도록 처리했습니다.
- 이 때 이미지가 캐시에 존재하는지 아닌지에 대한 체크는 Repository에서 요청을 할 때 하도록 구현했습니다.

## 🗓️ DateFormatter 구현

- 서버에서 받아온 “2023-04-06 14:22:00” 형태의 날짜 데이터를
요구사항인 “04/06(목) 14시” 형태로 변경하기 위해 DateFormatter 로직을 구현했습니다.
- 처음엔 아래처럼 DateFormatter 의 extension을 구현해서 처리해줬지만, 이 로직에서 몇가지 단점을 발견했습니다.
1. extension이라는 점
        - 특정 경우 (셀에 표시)에만 쓰이는 로직이기에, 
        모든 dateFormatter가 해당 메서드를 알 필요는 없을 듯 해보였음
2. 모든 cell이 매번 dateFormatter 인스턴스를 생성한다는 점
        - dateFormatter를 VC의 프로퍼티로 빼는 것으로 해결은 가능
3. 매번 DateFormatter의 dateFormat을 2번씩 변경 해주는 로직이 존재함
        - 그렇게 큰 비용이 아닐 수 있지만, CollectionView Cell 에 사용되는 로직이기에 아주 빠른 시간 내에 여러번 호출되는 성격을 지닌다고 판단함
    
    ```swift
    extension DateFormatter {
        func dateString(with dateString: String) -> String {
            self.dateFormat = "yyyy-MM-dd HH:mm:ss"
            guard let date = self.date(from: dateString) else { return "Formatting dateString fail"}
    
            self.dateFormat = "MM/dd(EEE) HH시"
            let formattedDate = self.string(from: date)
    
            return formattedDate
        }
    }
    ```
    
    ```swift
    func collectionView(_ collectionView: UICollectionView,
                        cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
    		...
        let date = DateFormatter().dateString(with: weather.dateString)
        cell.configure(date: date, temperature: temperature, iconCode: iconCode)
    		...
    }
    ```
    
- 위와 같은 문제를 해결하기 위해 이처럼 static 한 DateFormatter 인스턴스를 두개 생성 해 놓고, 해당 인스턴스를 사용하는 static func 를 구현하는 방식으로 Util을 만들었습니다.
    - 앱이 종료되기 전까지 계속해서 Cell을 표시하기 위해 사용되는 성격이기에 static이 적절하다고 생각했습니다.
    
    ```swift
    enum DateFormatUtil {
        
        // MARK: - Properties
        
        private static let convertFormatter: DateFormatter = {
            let formatter = DateFormatter()
            formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
            return formatter
        }()
        
        private static let resultFormatter: DateFormatter = {
            let formatter = DateFormatter()
            formatter.dateFormat = "MM/dd(EEE) HH시"
            return formatter
        }()
        
        // MARK: - Public
        
        static func format(with dateString: String) -> String {
            guard let date = convertFormatter.date(from: dateString) else {
                return "00/00(000) 00시"
            }
    
            let formattedDate = resultFormatter.string(from: date)
            return formattedDate
        }
    
    }
    ```
    
- 이 때 static 한 dateFormatter 인스턴스 두개를 계속해서 사용하는 방식이므로 Thread Safety가 걱정되긴 했지만, 공식문서의 DateFormatter는 Thread Safe 하다는 문구를 보고 해당 방식을 사용해도 문제없는 더 효율적인 로직이라 판단했습니다.
    
    > 공식문서: On iOS 7 and later `NSDateFormatter` is thread safe.
    > 

## 🎨 HeaderView만 따로 업데이트하게끔 구현

- 컬렉션뷰의 헤더뷰에서 필요로 하는 API 데이터는 현재 날씨 데이터이고, 
셀에서 필요로 하는 데이터는 주간 날씨 데이터여서 각각의 API 요청이 별도로 이루어졌습니다.
- 처음에는 헤더뷰를 업데이트 하기 위해서 CollectionView의 reloadData 해주었지만, 헤더뷰를 업데이트 하기 위해 전체 셀들을 업데이트 하는 것은 비효율적이란 생각이 들어 헤더뷰만 업데이트 하도록 리팩터링을 하였습니다.
    
    ```swift
    guard let headerView = self.collectionView.visibleSupplementaryViews(
        ofKind: UICollectionView.elementKindSectionHeader
    ).first as? WeatherHeaderView else { return }
    
    headerView.configure(
        with: currentWeather.weatherDetail,
        address: "\(placemark.locality ?? "") \(placemark.name ?? "")",
        icon: image
    )
    ```
    
- 하지만, 헤더를 제외한 “모든 셀들만” 업데이트 하는 로직에 대해서는 따로 해결하지 못하여 reloadData로 구현하였습니다.
    
    > 🤔 전체 셀들을 업데이트 할 때 헤더도 같이 업데이트 되는 로직이 큰 비효율은 아닌 듯 하다고 판단했지만, 혹시 더 좋은 방식이 있는지 궁금합니다!
    > 

## 💔 WeatherCell의 AutoLayout 충돌 해결 (Priority 설정)

- 현재 날씨를 나타내주는 CollectionView의 Header를 구현함에 있어서
기존의 flowLayout으로 되어있던 코드를 Compositional Layout의 list 방식으로 리팩터링 하면서, 
CompositionalLayout이 자동으로 header의 size를 지정하려고 하는 상황에서 디버그 콘솔창에 오류가 떴었고, 그것을 해결하는 과정이 있었습니다.
    
    > 충돌오류와 오류의 원인을 설명해주는 그림
    > 
    
    ```bash
    2023-04-06 14:03:13.617838+0900 WeatherForecast[71093:3166827] [LayoutConstraints] Unable to simultaneously satisfy constraints.
    	Probably at least one of the constraints in the following list is one you don't want. 
    	Try this: 
    		(1) look at each constraint and try to figure out which you don't expect; 
    		(2) find the code that added the unwanted constraint or constraints and fix it. 
    (
        "<NSLayoutConstraint:0x60000351ef30 WeatherForecast.WeatherHeaderView:0x15cc098a0.height == 100   (active)>",
        "<NSLayoutConstraint:0x600003513930 'UIView-Encapsulated-Layout-Height' WeatherForecast.WeatherHeaderView:0x15cc098a0.height == 17.6667   (active)>"
    )
    
    Will attempt to recover by breaking constraint 
    <NSLayoutConstraint:0x60000351ef30 WeatherForecast.WeatherHeaderView:0x15cc098a0.height == 100   (active)>
    
    Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
    The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKitCore/UIView.h> may also be helpful.
    ```
    
    <img width="982" alt="스크린샷 2023-04-06 오후 2 08 33" src="https://user-images.githubusercontent.com/59835351/230289879-37ebb338-a323-431b-a428-6a7a86f2e802.png">


- 사실 이게 완전한 해결이라기보다는 어쩌다 보니 해결 됐던 상황인데 처음 충돌 오류가 났을때 HeaderView의 Priority를 시험삼아 변경해 줬을 뿐인데 (낮췄는데) 해결이 돼버려서 그대로 진행했습니다.
- 하지만 사실 이렇게 HeaderView의 Priority를 변경하면서 예상했던 결과는 아래와 같았습니다.
    - 기본적으로 주어지는 Priority는 1000(max)이므로, 
    변경으로 인해 더 낮아진 HeaderView의 높이 Priority가 Priority변경을 하지 않았던 CompositionalLayout List에게 Constraint를 고수하기 위한 Priority 경쟁에서 밀려서 
    결국 충돌은 해결되지만, 더 우선순위가 높은 List의 높이로 변경 될 것이라는 결과를 예상했었지만 
    실행 결과는 예상과는 다르게 HeaderView가 원하던 높이(size 100) 으로 유지가 되고 있습니다. (우선순위가 더 낮음에도)
- 결과적으로는 해결된 듯 보이지만 저희 둘이서는 아무리 생각해보아도 해당 문제에 대한 확실한 원인과 해법을 찾을 수가 없었습니다.
- 현재는 `WeatherHeaderView` 의 `heightLayout.priority = .defaultHigh`이지만 해당 값을 어떤 다른 값으로 바꿔도 문제가 생기지 않는 미스테리한 상황입니다.😵‍💫

qwerty3345 committed:
design: Code UI 세팅

- SceneDelegate 삭제 (Scene의 개념 불필요)

sloDevz committed:
feat: CollectionView 선언 및 ReusableCell과 기본구현

sloDevz committed:
feat: Cell 요소 추가

qwerty3345 committed:
chore: 코드 정리, 폴더링 정리

qwerty3345 committed:
feat: 컬렉션뷰 Cell 내부 UI 설정

sloDevz committed:
feat: collectionView에 Header추가 및 배경이미지 지정

qwerty3345 committed:
refactor: 메인 ViewController 이름 변경

qwerty3345 committed:
refactor: 코드 정리, 네이밍 컨벤션 통일

sloDevz committed:
feat: header 내용 기본구현

sloDevz committed:
chore: 주석제거

sloDevz committed:
chore: 오타수정

qwerty3345 committed:
refactor: 코드 정리 (통일성 부여)

qwerty3345 committed:
Merge branch 'step3' of github.com:qwerty3345/ios-weather-forecast into step3

sloDevz committed:
refactor: CenterY Constraint 제거

qwerty3345 committed:
feat: API 섭씨로 받아오게 변경C

qwerty3345 committed:
feat: 컬렉션뷰 헤더에 API 데이터 나타나게 구현

qwerty3345 committed:
refactor: 현재날씨를 받아왔을 때 HeaderView만 Update되도록 변경

sloDevz committed:
weatherIconImageView CenterY 설정

qwerty3345 committed:
feat: 헤더뷰에 주소 정보 업데이트 하게끔 구현

qwerty3345 committed:
Merge remote-tracking branch 'refs/remotes/origin/step3'

sloDevz committed:
feat: DateFormatter extension에 날짜 형식 변경 메서드 구현

sloDevz committed:
feat: forecast 날짜 출력 구현, 기온출력 구현

qwerty3345 committed:
refactor: ViewController 상수값 분리

qwerty3345 committed:
refactor: Array index 안전하게 접근하게 extension 메서드 리팩터링

qwerty3345 committed:
refactor: DateFormatter 불필요한 인스턴스화 리팩터링

- 매번 DateFormat을 실행할 때 마다 불필요하게 DateFormatter를 인스턴스화 하던 로직을, static 형태로 Util화 함

qwerty3345 committed:
refactor: AddressManager의 completion을 optional 로 처리

- 현재 로직에서는 받아온 주소값을 Completion으로 넘기는 것이 아니라,
AddressManager 내부의 프로퍼티에 저장하므로, Completion을 필요할 때만 받아오게끔 변경함
- Optional parameter closure는 @escaping 성격이므로 키워드를 지움

sloDevz committed:
feat: 날씨 아이콘 출력 구현

qwerty3345 committed:
refactor: 이미지 URL 생성 로직 리팩터링

- 네이밍 변경, path 생성 하나의 함수로 분리

qwerty3345 committed:
fix: 셀 재사용 시 초기화 로직 구현

sloDevz committed:
feat: 이미지 캐싱 구현

sloDevz committed:
feat: 헤더 아이콘 출력 구현 및 각 Cell 업데이트 로직 변경

qwerty3345 committed:
refactor: updateListView로 네이밍 변경

qwerty3345 committed:
refactor: Compositional Layout 방식으로 변경

sloDevz committed:
feat: cell 컬러 투명으로 변경

qwerty3345 committed:
refactor: 컬렉션뷰 safearea 적용

qwerty3345 committed:
refactor: WeatherCell 상수값 분리

sloDevz committed:
feat: 당겨서 새로고침 구현

sloDevz committed:
merge: 머지

sloDevz committed:
refactor: 상수값 분리

qwerty3345 committed:
refactor: refresh control의 로딩 완료 시점 변경

- 첫번째 네트워킹이 완료된 시점으로 변경

qwerty3345 committed:
chore: string -> text 네이밍 변경

qwerty3345 committed:
chore: 네이밍 컨벤션 통일

qwerty3345 committed:
refactor: address 계산 속성으로 분리함으로서 쉽게 사용하게 변경

qwerty3345 committed:
merge: 오타 수정 커밋 머지

# Conflicts:
#	WeatherForecast/WeatherForecast/View/WeatherCell.swift
#	WeatherForecast/WeatherForecast/View/WeatherHeaderView.swift

wonhee009 commented on WeatherForecast/WeatherForecast/Controller/WeatherListViewController.swift:
```diff
@@ -0,0 +1,260 @@
+        setupLayout()
+        setupViewBackground()
+        setupLocationDataManager()
+        setupCollectionView()
```
해당 메서드 내에서 부르는 메서드들의 순서와 아래에 실제 메서드들이 정의되어 있는 순서가 상의한거 같은데요.
메서드 정의 순서의 기준이 따로 있는걸까요?
메서드를 어느 순서로 정의하느냐도 가독성에 영향을 준다고 생각합니다.

sloDevz replied:
지금껏 메서드 정의와 호출에대해 뚜렷한 기준이 없었지만 이번기회로 기준을 정하게 됐습니다.
메서드의 호출 순서는 최상단 뷰에 가까울 수록 먼저 호출하기로 기준을 잡았고,
정의 순서도 호출순서에 따라 상위에 위치하도록 했습니다!

wonhee009 commented on WeatherForecast/WeatherForecast/Controller/WeatherListViewController.swift:
```diff
@@ -0,0 +1,260 @@
+            updateListView()
+            endRefreshing()
+        }
+    }
```
두 데이터는 다른 타이밍에 setter 될 텐데요. 두 프로퍼티 모두 didSet에서 endRefreshing()을 호출해주고 있는거 같습니다.
endRefreshing 메서드 내에서 refreshControl이 active인지 아닌지에 대한 분기 처리는 필요 없을까요?

qwerty3345 replied:
DispatchGroup을 활용해서, 각각의 두 API 통신이 시작될 때 enter, 완료될때 leave 를 해주고 notify 에서 endRefreshing을 해줌으로서 두 API 통신이 모두 완료되었을 떄 업데이트 되도록 로직을 변경하였습니다!
또한 `isRefreshing` 체크 또한 처리하였습니다 :)

wonhee009 commented on WeatherForecast/WeatherForecast/Controller/WeatherListViewController.swift:
```diff
@@ -0,0 +1,260 @@
+            return UICollectionViewCell()
+        }
+
+        guard let weather = forecastDatas[safe: indexPath.row] else { return cell }
```
하나의 guard문으로 해결할 수 있지 않나요?

sloDevz replied:
수정했습니다!😊

wonhee009 commented on WeatherForecast/WeatherForecast/Manager/ImageCacheManager.swift:
```diff
@@ -0,0 +1,31 @@
+//
+
+import Foundation
+import UIKit
```
둘 다 import 해줘야 하나요?
apple 공식 문서의 Foundation, UIKit에 대해 학습해 보시는걸 추천합니다!

sloDevz replied:
수정했습니다 😉

qwerty3345 replied:
`UIKit`이 내부적으로 `Foundation`을 포함하고 있으므로, `UIKit` 만 남겼습니다!
또한 공부하는 과정에서 `CoreLocation`이 Foundation을 포함하고 있다는 것도 알게되어 중복된 import문을 제거 해 주었습니다 :)

wonhee009 commented on WeatherForecast/WeatherForecast/Networking/Repository/OpenWeatherRepository.swift:
```diff
@@ -6,6 +6,7 @@
 //
 
 import Foundation
+import UIKit
```
여기도 위의 코멘트와 동일합니다.

wonhee009 commented on WeatherForecast/WeatherForecast/Networking/Repository/OpenWeatherRepository.swift:
```diff
@@ -28,15 +29,19 @@ final class OpenWeatherRepository {
 
         static let weatherPath = "/data/2.5/weather"
         static let forecastPath = "/data/2.5/forecast"
+        static func iconImagePath(withID id: String) -> String { "/img/wn/\(id)@2x.png" }
```
이 친구는 메서드인거 같은데 여기에 위치한 이유가 있을까요?

qwerty3345 replied:
```swift
static let iconImagePath: String = "/img/wn/"
static let iconImageNamePostfix: String = "@2x.png"

...

urlComponents.path = Constant.iconImagePath + iconID + Constant.iconImageNamePostfix
```
처음엔 이렇게 분리시켜서 id값에 따라 합치는 식으로 구현했는데, 중간에 들어가는 id를 기준으로 저렇게 나누는 것이 애매해보여서 함수로 변경하였었습니다.

혹은 
```swift
static var iconImagePath: (String) -> String = { id in "/img/wn/\(id)@2x.png" }
...
urlComponents.path = Constant.iconImagePath(iconID)
```
이런 식의 구현도 가능 하다고 생각 했는데, 어쨌든 id값에 따라 변하는 값이기 떄문에 Constants로 간주하는 것은 적절하지 않을까요?
라자냐의 의견이 궁금합니다!

wonhee009 replied:
`computed property`를 활용할 수는 없나요?

qwerty3345 replied:
API Endpoint 를 생성해서, URL 을 생성하는 책임을 Repository에게서 분리하였습니다.
그러한 과정에서 enum의 case 와 computed property 를 활용 해 parameter 를 생성할 수 있도록 변경하였습니다!

wonhee009 commented on WeatherForecast/WeatherForecast/Util/DateFormatUtil.swift:
```diff
@@ -0,0 +1,37 @@
+    
+    // MARK: - Public
+    
+    static func format(with dateString: String) -> String {
+        guard let date = convertFormatter.date(from: dateString) else {
+            return "00/00(000) 00시"
+        }
+
+        let formattedDate = resultFormatter.string(from: date)
+        return formattedDate
+    }
```
`DateFormatter`를 이용해서 date로 변환하고, 그 date를 가지고 다시 format을 변환하는 로직으로 이해했는데요.
두번이나 Formatter를 이용해서 변환하는 이유가 있을까요?

qwerty3345 replied:
서버에서 받아온 값은 String으로 받아오기 떄문에, 
1. 해당 형식의 값을 Date로 변환하고, 
2. 해당 Date로 변환한 값을 원하는 형태로 화면에 나타내기 위해 String 으로 변환하는
2번의 과정이 필요하다 생각되었는데

2번 변환하지 않고 바로 서버에서 받아온 날짜 String을 화면에 나타내기 위한 형태의 String 으로 변환하는 방법이 존재할까요?

wonhee009 replied:
response를 디코딩할때 Date로 변환해서 디코딩하는 것도 방법일거 같습니다.

qwerty3345 replied:
서버에서 받아오는 데이터 중 timestamp 가 존재함을 뒤늦게 알아서, DateFormatter 를 한 번만 변환해도 되게 로직을 변경하였습니다!

wonhee009 commented on WeatherForecast/WeatherForecast/View/WeatherHeaderView.swift:
```diff
@@ -0,0 +1,141 @@
+    //MARK: - Constants
+
+    private enum Constants {
+        static let addressLabelSkeletonText: String = "서울특별시 용산새싹마을"
```
스켈레톤 텍스트로 이렇게 실제 내용을 넣어도 되나요?

sloDevz replied:
네 실제내용이 들어가면 안되는거지만 깜빡하고 바꾸지 못했던 부분입니다. 수정하겠습니다!

wonhee009 commented on WeatherForecast/WeatherForecast/View/WeatherHeaderView.swift:
```diff
@@ -0,0 +1,141 @@
+    private let weatherIconImageView: UIImageView = {
+        let view = UIImageView()
+        return view
+    }()
```
내부에서 따로 지정해 주는 부분이 없는거로 보이는데요. 
```suggestion
    private let weatherIconImageView = UIImageView()
```
이렇게 해도 되지 않나요?

sloDevz replied:
네 수정했습니다!

wonhee009 reviewed (changes requested):
수고하셨습니다:) 코멘트 남겨두었으니 확인 부탁드려요.

- 헤더를 제외하고 업데이트하기 위해서 ReloadItems와 같은 메서드를 사용할 수 있을거 같습니다.
- 설명해주신 AutoLayout 충돌에 대해서는 글만으로는 이슈 트래킹이 좀 어렵네요ㅠ

sloDevz committed:
refactor: 메서드 호출순서, 정의순서 변경

qwerty3345 committed:
fix: 두개의 API 작업이 모두 완료된 후에 Refreshing이 끝나도록 로직 변경

- DispatchGroup의 enter, leave, notify를 활용하여 리팩토링

sloDevz committed:
refactor: 불필요한 import 제거

sloDevz committed:
refactor: 스켈레톤 String 변경

qwerty3345 committed:
refactor: 뷰 생성 시 불필요 클로저 제거

qwerty3345 committed:
chore: 불필요 print문 제거

qwerty3345 committed:
refactor: 날짜 변환 DateFormatUtil 로직 변경

- String -> Date -> String 구조의 로직을
서버에서 받아온 값인 timestamp를 활용하여
Int -> Date -> String 으로 변경 (DateFormatter가 1개만 필요하게 변경)

qwerty3345 committed:
refactor: api 구성 변경 중

qwerty3345 committed:
refactor: Fetch Data 로직 추상화

qwerty3345 committed:
추상화 완료

qwerty3345 committed:
refactor: 불필요 구문 삭제

wonhee009 reviewed (approved):
수고하셨습니다~
